design-proposals/klip-36-grace-stream-stream-joins.md

```diff
@@ -0,0 +1,102 @@
+# KLIP 36 - GRACE period for stream-stream joins
+
+**Author**: agavra | 
+**Release Target**: 0.14 | 
+**Status**: In Discussion | 
+**Discussion**: TBD
+
+**tl;dr:** _Support controlling the GRACE period for a stream-stream join to improve disk space
+utilization for small windows_
+           
+## Motivation and background
+
+As described in https://github.com/confluentinc/ksql/issues/6152, the ability to specify
+the grace period for stream-stream joins allows more control over the disk space utilization.
+This is because a window needs to stay on disk until the grace period has elapsed in order to
+ensure that any events that come in timestamped during an old window can be processed as long
+as the grace period is active. This means that for windows dramatically smaller than the grace
+period (e.g. a minute) we need to keep that level of granularity for the default 24 hour window.
+
+## What is in scope
+
+- Syntax for specifying grace period within stream-stream joins
+
+## What is not in scope
+
+- N/A
+
+## Value/Return
+
+Enables onboarding use cases that perform stream-stream joins on high cardinality, high throughput
+datasets with small windows without exhausting too much disk space.
+
+## Public APIS
+
+Stream-stream joins require use of `WITHIN` clauses:
+
+```
+JOIN <stream> WITHIN <time unit> ON <condition>
+```
+
+This will be expanded to support a more complex `withinExpression` that leverages the same
+syntax as aggregation windows:
+
+```
+JOIN <stream> WITHIN (SIZE <time unit>, GRACE PERIOD <time unit>) ON <condition>
+```
+
+The old syntax will still be supported for backwards compatibility.
+
+## Design
+
+The `WITHIN` clause will be converted into a `KsqlWindowExpression`, and then converted into
+a `JoinWindows grace(final Duration afterWindowEnd)` call on the `JoinWindows` in the 
+`StreamStreamJoinBuilder`.
+
+## Test plan
+
+We will add the usual QTT tests to ensure that the system respects the new retention limits
+and manually test to ensure that we clean up the RocksDB state for expired windows.
+
+## LOEs and Delivery Milestones
+
+Small LOE - under 2 weeks of engineering time to deliver the syntax and implementation.
+
+## Documentation Updates
+
+The entry in `join-streams-and-tables.md` will be updated to include the following:
+
+```md
+When you join two streams, you must specify a WITHIN clause for matching
+records that both occur within a specified time interval and optionally a
```

Comment:
Ah, so it should be optionally?

Reply:
Curious to hear what other people think? Should GRACE PERIOD be optional? The current default of 24h is rather "random" and using a default of zero implies bad out-of-the-box behavior with regard to out-of-order data -- we would discard out-of-order records very aggressively as later records.

Reply:
Making it optional would follow the current pattern when you specify the grace period on a windowed aggregation.  As to the default of 24h, I'm not sure what the right answer here is.  For my specific use case of KSQL it's way too big, although for someone else it could be too small.  Ideally I'd like a init parameter I could set to have better control over the default behavior.  Then a 24h value  as a default is probably a good starting point.

Reply:
> Making it optional would follow the current pattern when you specify the grace period on a windowed aggregation.

Yes, but I am wondering if it should be mandatory for windowed aggregations, too? (Might be it's own KLIP thought.)

> As to the default of 24h, I'm not sure what the right answer here is.

I guess there is no right answer. This is exactly my point: if there is no reasonable default, it seems to imply that there shouldn't be any default but it should be mandatory?

Reply:
I think long term if should be required. However, we need to be backwards compatible, which means in the medium term it needs to be optional. 

Personally, I would change the wording from `and optionally `` to `and it is recommended to supply a`.

In the docs I'd also detail what the default it and explain what the trade of is here, i.e. longer grace period means more storage, shorter grace period means less storage required.

Reply:
I agree with @big-andy-coates, considering backward compatibility. We should recommend setting the grace period, but not require it for now.